File: JULY-2025/08-07-2025/p1751.scala
object Solution {
    def maxValue(events: Array[Array[Int]], k: Int): Int =
      loop(events.sorted(Ordering.by((e:Array[Int])=>(e(0), e(1)))), 0, k, 0, Array.ofDim[Int](k+1, events.length))

    def loop(events: Array[Array[Int]], index: Int, remainingEvents: Int, lastEventEndDay: Int, dp: Array[Array[Int]]): Int = 
      if (index >= events.length || remainingEvents == 0) 0
      else if (lastEventEndDay >= events(index)(0)) loop(events, index+1, remainingEvents, lastEventEndDay, dp)
      else if (dp(remainingEvents)(index) != 0) dp(remainingEvents)(index)
      else {
        dp(remainingEvents)(index) = math.max(
          loop(events, index+1, remainingEvents, lastEventEndDay, dp),
          loop(events, index+1, remainingEvents-1, events(index)(1), dp) + events(index)(2)
        )
        dp(remainingEvents)(index)
      }
}

File: JULY-2025/08-07-2025/p1751.dart
import 'dart:collection';
import 'dart:math';

class Solution {
  int maxValue(List<List<int>> events, int k) {
    events.sort((a, b) => a[1] - b[1]);

    final db = new SplayTreeMap<int, List<int>>();
    db[0] = List.filled(k + 1, 0);

    var res = 0;

    for (final event in events) {
      final end = event[1];
      final dp = List<int>.from(db[db.lastKeyBefore(end + 1)]!);

      final prev = db[db.lastKeyBefore(event[0])]!;
      
      for (int i = 0; i < k; i++)
        res = max(res, dp[i + 1] = max(dp[i + 1], prev[i] + event[2]));
      
      db[end] = dp;
    }

    return res;
  }
}